src/ring_theory/adjoin.lean

```diff
@@ -0,0 +1,166 @@
+      mv_polynomial.eval₂_X]; exact is_submonoid.mul_mem hp (subset_adjoin hn)))
+
+theorem adjoin_singleton_eq_range [decidable_eq R] [decidable_eq A] (x : A) :
+  adjoin R {x} = (polynomial.aeval R A x).range :=
+le_antisymm
+  (adjoin_le $ set.singleton_subset_iff.2 ⟨polynomial.X, polynomial.eval₂_X _ _⟩)
+  (λ y ⟨p, hp⟩, hp ▸ polynomial.induction_on p
+    (λ r, by rw [polynomial.aeval_def, polynomial.eval₂_C]; exact (adjoin R _).3 ⟨r, rfl⟩)
+    (λ p q hp hq, by rw alg_hom.map_add; exact is_add_submonoid.add_mem hp hq)
+    (λ n r ih, by rw [pow_succ', ← ring.mul_assoc, alg_hom.map_mul, polynomial.aeval_def _ _ _ polynomial.X,
+      polynomial.eval₂_X]; exact is_submonoid.mul_mem ih (subset_adjoin $ or.inl rfl)))
+
+theorem madjoin_union : (adjoin R (s ∪ t)).to_submodule =
+  (adjoin R s).to_submodule * (adjoin R t).to_submodule :=
+begin
+  rw [madjoin_eq_span, madjoin_eq_span, madjoin_eq_span, span_mul_span], congr' 1, ext z,
+  rw monoid.mem_closure_union_iff, split,
+  { rintro ⟨y, hys, z, hzt, rfl⟩, exact ⟨(_, _), ⟨hys, hzt⟩, rfl⟩ },
+  { rintro ⟨⟨y, z⟩, ⟨hys, hzt⟩, rfl⟩, exact ⟨_, hys, _, hzt, rfl⟩ }
+end
+variables {R s t}
+
+theorem adjoin_int (s : set R) : adjoin ℤ s = subalgebra_of_subring (ring.closure s) :=
+le_antisymm (adjoin_le subset_closure) (closure_subset subset_adjoin)
+
+theorem fg_trans (h1 : (adjoin R s).to_submodule.fg) (h2 : (adjoin (adjoin R s) t).to_submodule.fg) :
+  (adjoin R (s ∪ t)).to_submodule.fg :=
+begin
+  rcases fg_def.1 h1 with ⟨p, hp, hp'⟩,
+  rcases fg_def.1 h2 with ⟨q, hq, hq'⟩,
+  refine fg_def.2 ⟨set.image (λ z : A × A, z.1 * z.2) (p.prod q),
+    set.finite_image _ (set.finite_prod hp hq), le_antisymm _ _⟩,
+  { rw [span_le, set.image_subset_iff], rintros ⟨x, y⟩ ⟨hx, hy⟩,
+    change x * y ∈ _, refine is_submonoid.mul_mem _ _,
+    { have : x ∈ (adjoin R s).to_submodule, { rw ← hp', exact subset_span hx },
+      exact adjoin_mono (set.subset_union_left _ _) this },
+    have : y ∈ (adjoin (adjoin R s) t).to_submodule, { rw ← hq', exact subset_span hy },
+    change y ∈ adjoin R (s ∪ t), rw adjoin_union, exact this },
+  intros r hr, change r ∈ adjoin R (s ∪ t) at hr, rw adjoin_union at hr,
+  change r ∈ (adjoin (adjoin R s) t).to_submodule at hr,
+  rw [← hq', mem_span_iff_lc] at hr, rcases hr with ⟨l, hlq, rfl⟩,
+  haveI := classical.dec_eq A,
+  rw [lc.total_apply, finsupp.sum, mem_coe], refine sum_mem _ _,
+  intros z hz, change (l z).1 * _ ∈ _,
+  have : (l z).1 ∈ (adjoin R s).to_submodule := (l z).2,
+  rw [← hp', mem_span_iff_lc] at this, rcases this with ⟨l2, hlp, hl⟩, rw ← hl,
+  rw [lc.total_apply, finsupp.sum_mul], refine sum_mem _ _,
+  intros t ht, change _ * _ ∈ _, rw smul_mul_assoc, refine smul_mem _ _ _,
+  exact subset_span ⟨⟨t, z⟩, ⟨hlp ht, hlq hz⟩, rfl⟩
+end
+
+end algebra
+
+namespace subalgebra
+
+variables {R : Type u} {A : Type v}
+variables [comm_ring R] [comm_ring A] [algebra R A]
+
+def fg (S : subalgebra R A) : Prop :=
```

Comment:
Does it make sense to also have finitely generated algebras? The mathematician in me thinks they would be useful.

Reply:
This can be done in another PR